src/pages/home/report/withReportAndPrivateNotesOrNotFound.js

```diff
@@ -0,0 +1,129 @@
+import React, {useEffect, useMemo} from 'react';
+import PropTypes from 'prop-types';
+import _ from 'underscore';
+import lodashGet from 'lodash/get';
+import {withOnyx} from 'react-native-onyx';
+import getComponentDisplayName from '../../../libs/getComponentDisplayName';
+import reportPropTypes from '../../reportPropTypes';
+import withReportOrNotFound from './withReportOrNotFound';
+import * as Report from '../../../libs/actions/Report';
+import NotFoundPage from '../../ErrorPage/NotFoundPage';
+import FullScreenLoadingIndicator from '../../../components/FullscreenLoadingIndicator';
+import {withNetwork} from '../../../components/OnyxProvider';
+import compose from '../../../libs/compose';
+import * as ReportUtils from '../../../libs/ReportUtils';
+import networkPropTypes from '../../../components/networkPropTypes';
+import ONYXKEYS from '../../../ONYXKEYS';
+
+export default function (WrappedComponent) {
+    const propTypes = {
+        /** The HOC takes an optional ref as a prop and passes it as a ref to the wrapped component.
+         * That way, if a ref is passed to a component wrapped in the HOC, the ref is a reference to the wrapped component, not the HOC. */
+        forwardedRef: PropTypes.func,
+
+        /** The report currently being looked at */
+        report: reportPropTypes,
+
+        /** Information about the network */
+        network: networkPropTypes.isRequired,
+
+        /** Session of currently logged in user */
+        session: PropTypes.shape({
+            /** Currently logged in user accountID */
+            accountID: PropTypes.number,
+        }),
+
+        route: PropTypes.shape({
+            /** Params from the URL path */
+            params: PropTypes.shape({
+                /** reportID and accountID passed via route: /r/:reportID/notes */
+                reportID: PropTypes.string,
+                accountID: PropTypes.string,
+            }),
+        }).isRequired,
+    };
+
+    const defaultProps = {
+        forwardedRef: () => {},
+        report: {},
+        session: {
+            accountID: null,
+        },
+    };
+
+    // eslint-disable-next-line rulesdir/no-negated-variables
+    function WithReportAndPrivateNotesOrNotFound(props) {
+        const {route, report, network, session} = props;
```

Comment:
```suggestion
    function WithReportAndPrivateNotesOrNotFound({route, report, network, session}) {
```

Reply:
@ArekChr As this is HOC, I am passing the all the props to WrapperComponent so I need all of them:
```
        const rest = _.omit(props, ['forwardedRef']);
        return (
            <WrappedComponent
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...rest}
                ref={props.forwardedRef}
            />
        );
```

So I can't move `const {route, report, network, session} = props;` in component arguments.
What do you think?

Reply:
Now I understand your point of view. It makes sense. But it could be slightly improved, what do you think about spreading rest props here instead?

```
function WithReportAndPrivateNotesOrNotFound({ forwardedRef , ...props}) {
```

Then we will no need to use underscore omit and pass props with spread inside WrapperComponent

Reply:
This is a good idea in general but if I do it, what should i replace ` {...rest}` with? I want to pass all the properties to WrappedComponent as props (except forwardedRef). As we already spreaded props, I can't pass it to WrappedComponent anymore. Is there a solution for this?
```
            <WrappedComponent
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...rest}
                ref={forwardedRef}
            />
```

Reply:
I see what you are saying. 
To omit underscore usage, here is what I can do based on your suggestion:

**Option 1**
Changing to:
```
function WithReportAndPrivateNotesOrNotFound({ forwardedRef , ...rest}) {
.
.
<WrappedComponent
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={forwardedRef}
/>
```

In this case, I need to replace all the `route` usage with `rest.route`. I need to do the same with report, network and session objects. This is because `route` is not defined now. If I do this, its not that readable. 

**Option 2:**
Changing to:
```
function WithReportAndPrivateNotesOrNotFound({ forwardedRef , ...props}) {
.
.
<WrappedComponent
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      ref={forwardedRef}
/>
```

This is same as above. Better than option 1 in readability but I am repeating the `props.` everywhere now to access route, report, network and session. To avoid this, I originally used `const {route, report, network, session} = props;`. But I can go with this option if this is what you suggest is better.

Option 3:
Changing to:
```
function WithReportAndPrivateNotesOrNotFound({ forwardedRef, route, report, network, session, ...props}) {
.
.
<WrappedComponent
      // eslint-disable-next-line react/jsx-props-no-spreading
      {route, report, network, session, ...props}
      ref={forwardedRef}
/>
```

In this case, I am getting `route, report, network, session` in input arg and then getting it constructing it back in WrappedComponent. 

I see these 3 options to not use underscore and use `route, report, network, session` for my usecase and also pass it them to WrappedComponent along with other props. **Should I go with Option 2**?

By the way, to keep in mind for future usecases, can you tell me why we are avoiding `_.omit`? Should we skip it in general? Is it because it creates a shallow copy (still a copy) of the object?

Reply:
@salonikumawat28 Option 2 looks good to me. We can it combine with your current approach so you can destruct in props and below the props as it is now.

So it will look like this:
```
function WithReportAndPrivateNotesOrNotFound({ forwardedRef , ...props}) {
const {route, report, network, session} = props;

...

<WrappedComponent
      // eslint-disable-next-line react/jsx-props-no-spreading
      { ...props }
      ref={forwardedRef}
/>
```

Regarding underscore, this lib is removed while TypeScript migration so we will not use this anymore in future.

Reply:
Make the changes. Thanks.